File: entities.py
# .nttt
# .nwwai
import pygame
import pathfind
import utilityfuncs
from settings import *
import math

class Player:
    def __init__(self, x, y):
        self.x = x
        self.y = y
        self.real_x = x
        self.real_y = y
        self.dest_x = x
        self.dest_y = y
        self.hsp = 0
        self.vsp = 0
        self.move_path:list[tuple[int, int]] = list()
        self.surface = pygame.Surface((8, 8))
        self.surface.fill((255, 0, 0))
        self.move_speed = 2.5

    def calc_movement(self, gX, gY, tiles):
        self.dest_x = gX
        self.dest_y = gY
        self.move_path.clear()
        self.move_path = pathfind.pathfind(int(self.x / TILESIZE), int(self.y / TILESIZE), int(gX / TILESIZE), int(gY / TILESIZE), tiles)
        if not self.move_path:
            self.dest_x = self.x
            self.dest_y = self.y

    def movement(self):
        if len(self.move_path) == 0:
            if utilityfuncs.dist(self.dest_x, self.dest_y, self.real_x, self.real_y) > 2:
                dir_ = utilityfuncs.point_direction(self.real_x, self.real_y, self.dest_x, self.dest_y)
                self.real_x += math.cos(math.radians(dir_)) * self.move_speed
                self.real_y -= math.sin(math.radians(dir_)) * self.move_speed
            else:
                self.dest_x, self.dest_y = self.real_x, self.real_y
        else:
            x, y = (self.move_path[0][0] * TILESIZE + 16
                        , self.move_path[0][1] * TILESIZE + 16)

            if utilityfuncs.dist(self.real_x, self.real_y, x, y) > 5:
                dir_ = utilityfuncs.point_direction(self.x, self.y, x, y)
                self.front_dir = dir_
                self.real_x += math.cos(math.radians(dir_)) * self.move_speed
                self.real_y -= math.sin(math.radians(dir_)) * self.move_speed
            else:
                self.move_path.pop(0)
        self.x = pygame.math.lerp(self.x, self.real_x, 0.1)
        self.y = pygame.math.lerp(self.y, self.real_y, 0.1)

    def surf(self):
        return self.surface



class NPC:
    def __init__(self, x, y, surf: pygame.Surface, dialogues:list[str]):
        self.x = x
        self.y = y
        self.surface = surf

        self.dialogue:list[str] = dialogues
        self.post_dialogue_action:function = None

        self.real_x = x
        self.real_y = y
        self.dest_x = x
        self.dest_y = y
        self.hsp = 0
        self.vsp = 0
        self.move_path:list[tuple[int, int]] = list()
        # self.surface = pygame.Surface((8, 8))
        # self.surface.fill((255, 0, 0))
        self.move_speed = 2

    def calc_movement(self, gX, gY, tiles):
        self.dest_x = gX
        self.dest_y = gY
        self.move_path.clear()
        self.move_path = pathfind.pathfind(int(self.x / TILESIZE), int(self.y / TILESIZE), int(gX / TILESIZE), int(gY / TILESIZE), tiles)
        if not self.move_path:
            self.dest_x = self.x
            self.dest_y = self.y

    def movement(self):
        if len(self.move_path) == 0:
            if utilityfuncs.dist(self.dest_x, self.dest_y, self.real_x, self.real_y) > 2:
                dir_ = utilityfuncs.point_direction(self.real_x, self.real_y, self.dest_x, self.dest_y)
                self.real_x += math.cos(math.radians(dir_)) * self.move_speed
                self.real_y -= math.sin(math.radians(dir_)) * self.move_speed
            else:
                self.dest_x, self.dest_y = self.real_x, self.real_y
        else:
            x, y = (self.move_path[0][0] * TILESIZE + 16
                        , self.move_path[0][1] * TILESIZE + 16)

            if utilityfuncs.dist(self.real_x, self.real_y, x, y) > 5:
                dir_ = utilityfuncs.point_direction(self.x, self.y, x, y)
                self.front_dir = dir_
                self.real_x += math.cos(math.radians(dir_)) * self.move_speed
                self.real_y -= math.sin(math.radians(dir_)) * self.move_speed
            else:
                self.move_path.pop(0)
        self.x = pygame.math.lerp(self.x, self.real_x, 0.1)
        self.y = pygame.math.lerp(self.y, self.real_y, 0.1)

    def surf(self):
        return self.surface


class Prop:
    def __init__(self, x, y, z, surf: pygame.Surface):
        self.x = x
        self.y = y
        self.z = z
        self.surface = surf

    def surf(self):
        return self.surface

File: js/entities.js
// entities.js -- Player, NPC, AmbientNPC. Ported from entities.py.
import { pathfind } from "./pathfind.js";
import { dist, pointDirection, lerp, rad } from "./utils.js";
import { TILESIZE, PLAYER_MOVE_SPEED, NPC_MOVE_SPEED } from "./settings.js";

export class Player {
  constructor(x, y) {
    this.x = x; this.y = y;
    this.real_x = x; this.real_y = y;
    this.dest_x = x; this.dest_y = y;
    this.move_path = [];
    this.move_speed = PLAYER_MOVE_SPEED;
    this._frames = {};        // state -> [img]
    this._single = null;
    this.state = "idle";
    this.anim_fps = 6;
    this._anim_t = 0;
  }
  set_sprite(s) { this._single = s; }
  set_frames(f) { this._frames = f; }

  _get_frame() {
    const frames = this._frames[this.state] || this._frames["idle"];
    if (frames && frames.length) {
      const idx = Math.floor(this._anim_t * this.anim_fps) % frames.length;
      return frames[idx];
    }
    return null;
  }
  calc_movement(gx, gy, tiles) {
    this.dest_x = gx; this.dest_y = gy;
    this.move_path = pathfind(
      Math.floor(this.x / TILESIZE), Math.floor(this.y / TILESIZE),
      Math.floor(gx / TILESIZE), Math.floor(gy / TILESIZE), tiles);
    if (!this.move_path.length) { this.dest_x = this.x; this.dest_y = this.y; }
  }
  move_to(gx, gy, tiles) { this.calc_movement(gx, gy, tiles); }
  has_arrived() { return this.move_path.length === 0; }
  is_moving() { return this.move_path.length > 0; }

  movement() {
    if (this.move_path.length === 0) {
      this.dest_x = this.real_x; this.dest_y = this.real_y;
    } else {
      const [tx, ty] = this.move_path[0];
      const px = tx * TILESIZE + 16, py = ty * TILESIZE + 16;
      if (dist(this.real_x, this.real_y, px, py) > 5) {
        const d = pointDirection(this.x, this.y, px, py);
        this.real_x += Math.cos(rad(d)) * this.move_speed;
        this.real_y -= Math.sin(rad(d)) * this.move_speed;
      } else {
        this.move_path.shift();
      }
    }
    this.x = lerp(this.x, this.real_x, 0.1);
    this.y = lerp(this.y, this.real_y, 0.1);
    if (this.is_moving()) { this.state = "walk"; }
    else if (this.state === "walk") { this.state = "idle"; }
    this._anim_t += 1 / 60;
  }
  bob_offset() {
    if ((this._single === null && !Object.keys(this._frames).length) || !this.is_moving()) return 0;
    return Math.sin(this._anim_t * 8) * 3;
  }
  surf() {
    const f = this._get_frame();
    if (f) return f;
    if (this._single) return this._single;
    return null; // invisible until a role is chosen
  }
}

export class NPC {
  constructor(x, y, surface) {
    this.x = x; this.y = y;
    this.real_x = x; this.real_y = y;
    this.move_path = [];
    this.surface = surface;
    this.move_speed = NPC_MOVE_SPEED;
  }
  movement() {
    if (this.move_path.length === 0) return;
    const [tx, ty] = this.move_path[0];
    const px = tx * TILESIZE + 16, py = ty * TILESIZE + 16;
    if (dist(this.real_x, this.real_y, px, py) > 5) {
      const d = pointDirection(this.x, this.y, px, py);
      this.real_x += Math.cos(rad(d)) * this.move_speed;
      this.real_y -= Math.sin(rad(d)) * this.move_speed;
    } else { this.move_path.shift(); }
    this.x = lerp(this.x, this.real_x, 0.1);
    this.y = lerp(this.y, this.real_y, 0.1);
  }
  surf() { return this.surface; }
}

export class AmbientNPC {
  constructor(x, y, surface) { this.x = x; this.y = y; this.surface = surface; }
  surf() { return this.surface; }
}
